apollo-federation/src/sources/connect/json_selection/parser.rs

```diff
@@ -192,23 +183,65 @@ impl NamedSelection {
     }
 }
 
-// PathSelection ::= (VarPath | KeyPath) SubSelection?
+// PathSelection ::= (VarPath | KeyPath | AtPath) SubSelection?
 // VarPath       ::= "$" (NO_SPACE Identifier)? PathStep*
 // KeyPath       ::= Key PathStep+
+// AtPath        ::= "@" PathStep*
 // PathStep      ::= "." Key | "->" Identifier MethodArgs?
 
 #[derive(Debug, PartialEq, Clone, Serialize)]
-pub enum PathSelection {
+pub struct PathSelection {
+    pub(super) path: PathList,
+}
+
+impl PathSelection {
+    pub fn parse(input: &str) -> IResult<&str, Self> {
+        let (input, path) = PathList::parse(input)?;
+        Ok((input, Self { path }))
+    }
+
+    pub fn is_single_key(&self) -> bool {
+        self.path.is_single_key()
+    }
+
+    pub fn from_slice(keys: &[Key], selection: Option<SubSelection>) -> Self {
+        Self {
+            path: PathList::from_slice(keys, selection),
+        }
+    }
+
+    pub fn collect_paths(&self) -> Vec<Key> {
+        self.path.collect_paths()
+    }
+
+    pub fn next_subselection(&self) -> Option<&SubSelection> {
+        self.path.next_subselection()
+    }
+
+    pub fn next_mut_subselection(&mut self) -> Option<&mut SubSelection> {
+        self.path.next_mut_subselection()
+    }
+}
+
+impl From<PathList> for PathSelection {
+    fn from(path: PathList) -> Self {
+        Self { path }
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize)]
+pub(super) enum PathList {
```

Comment:
Can you add some docstrings to both this and each variant describing concretely what they mean? Like, for `Var(String, Box<PathList>)`, corresponds to `$this.something.somethingElse` where `.0` contains `$this`, and `.1` is the rest of the path (if that's true...)

Reply:
https://github.com/apollographql/router/pull/5762/commits/c57b20dfd9cf7e41dbdf4ebd8129f5f33e6a1686